lib/policy_machine_storage_adapters/active_record.rb

```diff
@@ -482,6 +476,96 @@ def operations
       end
     end # End of POLICY_ELEMENT_TYPES iteration
 
+    # Given a policy element class and a set of conditions, returns an
+    # ActiveRecord_Relation with those conditions applied
+    def build_active_record_relation(pe_class:, conditions:, ignore_case:)
+      # If any condition is case-insensitive, the nodes need to be built
+      # individually with Arel.
+      if ignore_case
+        conditions.map do |k, v|
+          # Arel matches provides agnostic case insensitive sql for MySQL and
+          # Postgres. This should always evaluate to an ActiveRecord_Relation.
+          if ignore_case_applies?(ignore_case, k)
+            build_arel_insensitive(pe_class: pe_class, key: k, value: v)
+          else
+            build_arel_sensitive(pe_class: pe_class, key: k, value: v)
+          end
+        # Reduce Arel nodes into ActiveRecord_Relation
+        end.reduce(pe_class.where(nil)) { |rel, e| rel.where(e) }
+      else
+        # If all conditions are case-sensitive, a direct where call can be used.
+        pe_class.where(conditions.to_h)
+      end
+    end
+
+    # Build Arel nodes using case-insensitive matching
+    def build_arel_insensitive(pe_class:, key:, value:)
+      unless value.is_a?(Array)
+        pe_class.arel_table[key].matches(value)
+      else
+        pe_class.arel_table[key].matches_any(value)
+      end
+    end
+
+    # Build Arel nodes using case-sensitive equality checking
+    def build_arel_sensitive(pe_class:, key:, value:)
+      unless value.is_a?(Array)
+        pe_class.arel_table[key].eq(value)
+      else
+        unless value.empty?
+          pe_class.arel_table[key].eq_any(value)
+        else
+          # Arel blows up with empty array passed to eq_any
+          # See: https://github.com/rails/arel/issues/368
+          ::Arel::Nodes::SqlLiteral.new("(NULL)")
+        end
+      end
+    end
+
+    # Iterate over the input scope and return elements that match the extra
+    # attribute conditions
+    def filter_by_extra_attributes(scope:, extra_attribute_conditions:, pe_class:, ignore_case:)
+      unless extra_attribute_conditions.empty?
+        ids = scope.select do |pe|
```

Comment:
Looks like we're iterating over this scope twice – once for the select and then once again for the map at the end. If we use a reduce here instead then we can cut down our number of iterations over the scope to 1.

Reply:
Excellent catch!